src/intel_pmu.c

```diff
@@ -401,12 +406,6 @@ static int pmu_add_hw_events(struct eventlist *el, char **e, size_t count) {
     char *s, *tmp;
     for (s = strtok_r(events, ",", &tmp); s; s = strtok_r(NULL, ",", &tmp)) {
 
-      /* Multiple events parsed in one entry */
-      if (group_events_count == 1) {
-        /* Mark previously added event as group leader */
-        el->eventlist_last->group_leader = 1;
-      }
-
       /* Allocate memory for event struct that contains array of efd structs
          for all cores */
       struct event *e =
```

Comment:
This reuse of 'e' variable breaks my brain )

Maybe somebody will make code clearer? I'm unable to compile this plugin currently to make proper checks.

Thanks!

Reply:
http://docs.opnfv.org/en/latest/submodules/barometer/docs/release/userguide/feature.userguide.html#intel-pmu-plugin should have some instructions to help with compilation.

Reply:
```
root@xxx:~/xxx/pmu-tools-107/jevents# make
make -C examples
make[1]: Entering directory `/xxx/xxx/pmu-tools-107/jevents/examples'
cc -L ..  rtest.o  -ljevents -o rtest
../libjevents.a(rdpmc.o): In function `rdpmc_read':
/xxx/xxx/pmu-tools-107/jevents/rdpmc.c:135: undefined reference to `__builtin_ia32_rdpmc'
```

Which compiler or library / version should I use? Something is outdated in my playground, I think.  ;-)

Reply:
Ok, replaced the sand with more fresh one )

Can you help me ?

```
[2018-02-10 06:40:06] plugin_load: plugin "intel_pmu" successfully loaded.
[2018-02-10 06:40:06] intel_pmu: pmu_config:239
[2018-02-10 06:40:06] intel_pmu: Config:
[2018-02-10 06:40:06] intel_pmu:   hw_cache_events   : 1
[2018-02-10 06:40:06] intel_pmu:   kernel_pmu_events : 1
[2018-02-10 06:40:06] intel_pmu:   software_events   : 1
[2018-02-10 06:40:06] intel_pmu:   event_list_file   : /root/.cache/pmu-events/GenuineIntel-6-3A-core.json
[2018-02-10 06:40:06] intel_pmu:   hardware_events[0]: L2_RQSTS.CODE_RD_HIT,L2_RQSTS.CODE_RD_MISS
[2018-02-10 06:40:06] intel_pmu:   hardware_events[1]: L2_RQSTS.ALL_CODE_RD
[2018-02-10 06:40:06] Done parsing `/opt/collectd/share/collectd/types.db'
[2018-02-10 06:40:06] interval_g = 10.000;
[2018-02-10 06:40:06] timeout_g = 2;
[2018-02-10 06:40:06] hostname_g = test;
[2018-02-10 06:40:06] intel_pmu: pmu_init:498
[2018-02-10 06:40:06] intel_pmu: Cannot resolve L2_RQSTS.CODE_RD_HIT
[2018-02-10 06:40:06] intel_pmu: Cannot resolve L2_RQSTS.CODE_RD_MISS
[2018-02-10 06:40:06] intel_pmu: Cannot resolve L2_RQSTS.ALL_CODE_RD
```

Why these all `Cannot resolve` happen? All these events are in 'event_list_file':

```
# grep -E "L2_RQSTS.CODE_RD_HIT|L2_RQSTS.CODE_RD_MISS|L2_RQSTS.ALL_CODE_RD" /root/.cache/pmu-events/GenuineIntel-6-3A-core.json
    "EventName": "L2_RQSTS.CODE_RD_HIT",
    "EventName": "L2_RQSTS.CODE_RD_MISS",
    "EventName": "L2_RQSTS.ALL_CODE_RD",
```

What I'm missing? I'm a newbie here )

Reply:
I guess it can be wrong json for your CPU. Did you download one using script
`pmu-tools/event_download.py` ?

For some reason it's not described in the guide.
I would also try to put events file not in root directory, but it shouldn't matter.

Reply:
Yes, that file was downloaded by `pmu-tools/event_download.py`.
Will that work on Xen VM?

I thought that will fail on read, not on init phase.

Reply:
Using chroot environment may have effect on ability to read pmu events.
What about other events like KernelPMUEvents and SoftwareEvents, do they work fine?

To investigate it further and help you I would need more details about your environment, pmu plugin configuration and possibly more debug output from collectd. Maybe there are more tips in debug logs.
To make logs clearer you can disable all other events beside hardware events.
```
ReportHardwareCacheEvents false
ReportKernelPMUEvents false
ReportSoftwareEvents false
```

Note that setting in /proc/sys/kernel/perf_event_paranoid affect the rights to read perf events.
And last maybe stupid question, do you have perf tool available and working?
Try for example `perf stat ls` and check the output.

Reply:
>What about other events like KernelPMUEvents and SoftwareEvents, do they work fine?

These send some metrics (checked each independently).

pmu plugin configuration based on manpage:

```
LoadPlugin intel_pmu
<Plugin intel_pmu>
       ReportHardwareCacheEvents true
       ReportKernelPMUEvents false
       ReportSoftwareEvents false
       EventList "/root/.cache/pmu-events/GenuineIntel-6-3A-core.json"
       HardwareEvents "L2_RQSTS.CODE_RD_HIT,L2_RQSTS.CODE_RD_MISS" "L2_RQSTS.ALL_CODE_RD"
</Plugin>
```

After something changed on system (possible due to /sys mounted inside chroot), the output is changed too - `Cannot parse qualifier umask` appears:

```
[2018-02-10 15:24:06] plugin_load: plugin "intel_pmu" successfully loaded.
[2018-02-10 15:24:06] intel_pmu: pmu_config:239
[2018-02-10 15:24:06] intel_pmu: Config:
[2018-02-10 15:24:06] intel_pmu:   hw_cache_events   : 1
[2018-02-10 15:24:06] intel_pmu:   kernel_pmu_events : 0
[2018-02-10 15:24:06] intel_pmu:   software_events   : 0
[2018-02-10 15:24:06] intel_pmu:   event_list_file   : /root/.cache/pmu-events/GenuineIntel-6-3A-core.json
[2018-02-10 15:24:06] intel_pmu:   hardware_events[0]: L2_RQSTS.CODE_RD_HIT,L2_RQSTS.CODE_RD_MISS
[2018-02-10 15:24:06] intel_pmu:   hardware_events[1]: L2_RQSTS.ALL_CODE_RD
[2018-02-10 15:24:06] Done parsing `/opt/collectd/share/collectd/types.db'
[2018-02-10 15:24:06] interval_g = 10.000;
[2018-02-10 15:24:06] timeout_g = 2;
[2018-02-10 15:24:06] hostname_g = test;
[2018-02-10 15:24:06] intel_pmu: pmu_init:498
Cannot parse qualifier umask
[2018-02-10 15:24:06] intel_pmu: Cannot resolve L2_RQSTS.CODE_RD_HIT
Cannot parse qualifier umask
[2018-02-10 15:24:06] intel_pmu: Cannot resolve L2_RQSTS.CODE_RD_MISS
Cannot parse qualifier umask
[2018-02-10 15:24:06] intel_pmu: Cannot resolve L2_RQSTS.ALL_CODE_RD
[2018-02-10 15:24:06] intel_pmu: Events:
[2018-02-10 15:24:06] intel_pmu:   event       : L1-dcache-loads
[2018-02-10 15:24:06] intel_pmu:     group_lead: 0
[2018-02-10 15:24:06] intel_pmu:     end_group : 0
[2018-02-10 15:24:06] intel_pmu:     type      : 0x3
[2018-02-10 15:24:06] intel_pmu:     config    : 0
[2018-02-10 15:24:06] intel_pmu:     size      : 64
[2018-02-10 15:24:06] intel_pmu:   event       : L1-dcache-load-misses
[2018-02-10 15:24:06] intel_pmu:     group_lead: 0
[2018-02-10 15:24:06] intel_pmu:     end_group : 0
[2018-02-10 15:24:06] intel_pmu:     type      : 0x3
[2018-02-10 15:24:06] intel_pmu:     config    : 0x10000
[2018-02-10 15:24:06] intel_pmu:     size      : 64
[...cut..]
Cannot open perf event for L1-dcache-loads/0: No such file or directory
[2018-02-10 15:24:06] intel_pmu: perf event 'L1-dcache-loads' is not available (cpu=0).
Cannot open perf event for L1-dcache-loads/1: No such file or directory
[2018-02-10 15:24:06] intel_pmu: perf event 'L1-dcache-loads' is not available (cpu=1).
Cannot open perf event for L1-dcache-loads/2: No such file or directory
```

>Note that setting in /proc/sys/kernel/perf_event_paranoid affect the rights to read perf events.

Was "1", currently set to "0".

>And last maybe stupid question, do you have perf tool available and working?
>Try for example perf stat ls and check the output.

That is not a stupid question ;-) As I say, I'm newbie here. 

`perf stat` output is interesting. I have installed two versions from Debian packages.
Their output differs.

```
root@test01:/usr/bin# dpkg-query -S /usr/bin/perf_3.16
linux-tools-3.16: /usr/bin/perf_3.16

root@test01:/usr/bin# dpkg-query -S /usr/bin/perf_4.9
linux-perf-4.9: /usr/bin/perf_4.9

root@test01:/tmp/empty# perf_3.16 stat ls

 Performance counter stats for 'ls':

          0.988017      task-clock (msec)         #    0.721 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
               256      page-faults               #    0.259 M/sec
         3,437,961      cycles                    #    3.480 GHz
   <not supported>      stalled-cycles-frontend
   <not supported>      stalled-cycles-backend
         1,162,296      instructions              #    0.34  insns per cycle
           268,520      branches                  #  271.777 M/sec
            10,606      branch-misses             #    3.95% of all branches

       0.001369435 seconds time elapsed

root@test01:/tmp/empty# perf_4.9 stat ls

 Performance counter stats for 'ls':

   <not supported>      task-clock
   <not supported>      context-switches
   <not supported>      cpu-migrations
   <not supported>      page-faults
   <not supported>      cycles
   <not supported>      instructions
   <not supported>      branches
   <not supported>      branch-misses

       0.001645428 seconds time elapsed
```

Reply:
OK, we have something now :)
The `Cannot parse qualifier umask` means that file `/sys/devices/<cpu|uncore>/format/umask` can't be read for any reason.
Check your access and if it exists.

Perf should fit the kernel version. Do you use debian with 3.16 kernel? Than it is the right version for you. What it the default (the one for perf without modifiers) `perf --version`?

To check manually if event can be read use command:
`perf stat -e r214 -a sleep 2`
Where in this case 14 is "EventCode" and 2 is "UMask". You can find the values in json file for particular event.
Generally you need a coherence of kernel, perf and free access to /sys to get it work.

Btw. the `ReportHardwareCacheEvents` is not necessary to read `HardwareEvents`.

Reply:
Thanks, now it works, even in chroot.